proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,235 @@
+`RewardHistoryEntry` will be added to the `EpochRewardHistory` array with the
+total amount of reward, a distributed value equal to the sum of the vote
+account rewards, and the number of reward reserve accounts for stake reward
+distribution. If the length of the history array reaches the maximum depth (512), the
+oldest entry will be removed from the array (i.e., "popped").
+
+
+### `EpochRewardReserve` sysvar account
+
+A total of 1,000 sysvar addresses are reserved for `SysvarRewardReserve`
+accounts.
+
+```
+crate::declare_sysvar_id!("SysvarRewardReserve00111111111111111111111111111111", EpochRewardReserve);
+crate::declare_sysvar_id!("SysvarRewardReserve00211111111111111111111111111111", EpochRewardReserve);
+crate::declare_sysvar_id!("SysvarRewardReserve00311111111111111111111111111111", EpochRewardReserve);
+...
+crate::declare_sysvar_id!("SysvarRewardReserve99911111111111111111111111111111", EpochRewardReserve);
+
+struct EpochRewardReserve {
+   balance_in_lamport: u64,
+   rewards_hash: Hash,
+}
+```
+
+Each `RewardsReserve` contains the total balance (in lamports) and the hash for
+all rewards that will be distributed from it.
+
+```
+rewards_hash = hash([(stake_account_key, reward_amount)])
+```
+
+At the end of reward computation phase, M `EpochRewardsReserve` sysvar
+accounts are updated, where `M` corresponds to how many blocks will be in
+the reward distribution phase, which is described in the next section.
+
+
+### Reward Distribution
+
+Reward distribution phase happens right after reward computation phase and
+lasts for `M` blocks.
+
+Based on current validator client's performance, 64 additional accounts can be
+stored per tick time without much performance impact for the overall block
+time. Given that each block time is 64 ticks time long, a total of 4,096
+accounts can be distributed per block. The total number of blocks needed tto
+distributed rewards is
+
+```
+M = ((4096 - 1)+num_stake_accounts)/4096
+```
+
+During the reward distribution phase, rewards will be distributed from the
+`EpochRewardReserve` sysvar account with the address
+`SysvarRewardReserve_{i_th}_111111111` for the `i_th` block. As the
+stake_accounts receive their rewards from the reserve account, the balance of
+the EpochRewardReserve will decrease. Once all rewards have been distributed, the
+balance of the EpochRewardReserve account should be reduced to 0. Before each
+reward distribution, the `EpochRewardReserve` account's balance is check to
+make sure there is enough balance to distribute the reward. After a reward is
+distributed, the balance in `EpochRewardReserve` account is decreased by the
+amount of the reward that are distributed. The hash of reward_account pubkey and reward amount are accumulated when rewards are paying out.
+After all the rewards are
+distributed from the reserve account for the block, the hash computed from all the rewards
+distribution is verified against the hash stored in the `EpochReserve` to make sure
```

Comment:
By `EpochReserve`, do you mean `RewardHistoryEntry`?  The root hash makes sense to me now, thank for the explanation here, but I don't see it in `RewardHistoryEntry` anymore?